presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/RuleStore.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Streams;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class RuleStore
+{
+    private final ListMultimap<Class<? extends PlanNode>, Rule> rulesByClass;
+
+    private RuleStore(ListMultimap<Class<? extends PlanNode>, Rule> rulesByClass)
+    {
+        this.rulesByClass = ImmutableListMultimap.copyOf(rulesByClass);
+    }
+
+    public Stream<Rule> getCandidates(PlanNode planNode)
+    {
+        return Streams.stream(superPlanNodes(planNode.getClass()))
+                .flatMap(clazz -> rulesByClass.get(clazz).stream());
+    }
+
+    private static Iterator<Class<? extends PlanNode>> superPlanNodes(Class<? extends PlanNode> planNodeClass)
```

Comment:
Also, I think this would make the code a bit more readable:

```java
return new AbstractIterator<Class<? extends PlanNode>>() {
    private Class<?> current = clazz;
    
    @Override
    protected Class<? extends PlanNode> computeNext()
    {
        if (!PlanNode.class.isAssignableFrom(clazz)) {
            return endOfData();
        }

        Class<? extends PlanNode> result = (Class<? extends PlanNode>) current;
        current = current.getSuperclass();

        return result;
    }
};
```